test/xskbench/km/sys/driver.c

```diff
@@ -30,237 +28,153 @@ static BOOLEAN IsSessionActive;
-    (PNPI_CLIENT_CLEANUP_BINDING_CONTEXT_FN)XbDrvNmrCleanupXdpApiBindingContext,
-    {
-        0, // Version
-        sizeof(NPI_REGISTRATION_INSTANCE),
-        &NPI_XDPAPI_INTERFACE_ID,
-        &NPI_XBDRV_MODULEID,
-        XDP_API_VERSION_1, // Number
-        NULL // NpiSpecificCharacteristics
-    } // ClientRegistrationInstance
-};
-
-XDP_API_PROVIDER_BINDING_CONTEXT *
-CxPlatXdpApiGetProviderBindingContext(
-    VOID
+XDP_STATUS
+CxPlatXdpCreateProgram(
+    _In_ UINT32 InterfaceIndex,
+    _In_ const XDP_HOOK_ID *HookId,
+    _In_ UINT32 QueueId,
+    _In_ XDP_CREATE_PROGRAM_FLAGS Flags,
+    _In_reads_(RuleCount) const XDP_RULE *Rules,
+    _In_ UINT32 RuleCount,
+    _Out_ HANDLE *Program
     )
 {
-    return XdpApiProviderBindingContext;
+    return XdpApi->XdpCreateProgram(
+        XdpApiProviderBindingContext,
+        InterfaceIndex,
+        HookId,
+        QueueId,
+        Flags,
+        Rules,
+        RuleCount,
+        Program);
 }
 
 VOID
-CxPlatXdpApiInitialize(
-    VOID
+CxPlatPrintStats(
+    MY_QUEUE *Queue
     )
 {
-    NTSTATUS Status;
-
-    TraceEnter(TRACE_CONTROL, "-");
-
-    Status =
-        NmrRegisterClient(
-            &XbDrvNmrXdpApiClientCharacteristics,
-            NULL,
-            &NmrRegistrationHandle);
-    if (!NT_SUCCESS(Status)) {
-        goto Done;
+    KFLOATING_SAVE FloatingSave;
+    if (KeSaveFloatingPointState(&FloatingSave) == STATUS_SUCCESS) {
+        PrintFinalStats(Queue);
+        KeRestoreFloatingPointState(&FloatingSave);
     }
-
-    KeWaitForSingleObject(&BoundToProvider, Executive, KernelMode, FALSE, NULL);
-
-Done:
-
-    TraceExitStatus(TRACE_CONTROL);
-
-    return;
 }
 
 VOID
-CxPlatXdpApiUninitialize(
-    VOID
+CxPlatQueueCleanup(
+    MY_QUEUE *Queue
     )
 {
-    NTSTATUS Status = STATUS_SUCCESS;
-
-    TraceEnter(TRACE_CONTROL, "-");
-
-    if (NmrRegistrationHandle != NULL) {
-        Status = NmrDeregisterClient(NmrRegistrationHandle);
-        ASSERT(Status == STATUS_PENDING);
-
-        if (Status == STATUS_PENDING) {
-            Status = NmrWaitForClientDeregisterComplete(NmrRegistrationHandle);
-            ASSERT(Status == STATUS_SUCCESS);
-        }
+    if (Queue->rxProgram != NULL) {
+        XdpApi->XdpCloseHandle(Queue->rxProgram);
+        Queue->rxProgram = NULL;
+    }
+    if (Queue->sock != NULL) {
+        XdpApi->XdpCloseHandle(Queue->sock);
+        Queue->sock = NULL;
+    }
 
-        NmrRegistrationHandle = NULL;
+    if (Queue->umemReg.Address != NULL) {
+        CXPLAT_VIRTUAL_FREE(Queue->umemReg.Address, 0, MEM_RELEASE, POOLTAG_UMEM);
+        Queue->umemReg.Address = NULL;
     }
+    if (Queue->freeRingLayout != NULL) {
+        CXPLAT_FREE(Queue->freeRingLayout, POOLTAG_FREERING);
+        Queue->freeRingLayout = NULL;
+    }
+}
 
-    TraceExitStatus(TRACE_CONTROL);
+BOOLEAN
+CxPlatEnableLargePages(
+    VOID
+    )
+{
+    return TRUE;
+}
 
-    return;
+VOID
+CxPlatAlignMemory(
```

Comment:
i'm not sure this should be a no-op? large pages are still interesting in kernel mode, and have similar constraints

Reply:
you can use LARGE_PAGE_SIZE in kernel mode instead of GetLargePageSize

Reply:
@nigriMSFT Any idea?

Reply:
Sgtm. The no-op for kernel mode was simply a prototype shortcut

Reply:
@nigriMSFT 
> you can use LARGE_PAGE_SIZE in kernel mode instead of GetLargePageSize

You mean `GetLargePage**Minimun**`?
And there is no LARGE_PAGE_SIZE. Is there any header to get the value? or You mean I should define by myself? if so 2MB?

Reply:
oh, you're right, it's defined in a non-public Windows header. i'm fine leaving this as-is for now.